doc: 2 guides on modernizing apps: for apps migrated to 2211.19 and for apps migrated to Feb 2025 release (#19924)
SAP/spartacus: docs/migration/2211_35/modernize-apps-migrated-from-6.8-to-2211.19.md

-        "prerender": {
-          "builder": "@angular-devkit/build-angular:prerender",
-          "options": {
-            "routes": [
-              "/"
-            ]
-          },
-          "configurations": {
-            "production": {
-              "browserTarget": "YOUR-APP-NAME:build:production",
-              "serverTarget": "YOUR-APP-NAME:server:production"
-            },
-            "development": {
-              "browserTarget": "YOUR-APP-NAME:build:development",
-              "serverTarget": "YOUR-APP-NAME:server:development"
-            }
-          },
-          "defaultConfiguration": "production"
-        }
```

### `package.json`

Please change the following `"scripts"` properties (because the new `application` builder handles the SSR and prerendering with different commands):

1. Please remove properties `"dev:ssr"` and `"prerender"`

```diff
   "scripts": {
-    "dev:ssr": "ng run YOUR-APP-NAME:serve-ssr",
-    "prerender": "ng run YOUR-APP-NAME:prerender"
```

2. Please change value of the property `"build:ssr"` to `"ng build"`

```diff
   "scripts": {
-     "build:ssr": "ng build && ng run YOUR-APP-NAME:server",
+     "build:ssr": "ng build"
```

> 3. Please change the value of the property `"serve:ssr"` to `"node dist/YOUR-APP-NAME/server/server.mjs"`

Author: ```suggestion
3. Please rename the property `"serve:ssr"` to `"serve:ssr:YOUR-APP-NAME"` and change its value to `"node dist/YOUR-APP-NAME/server/server.mjs"`
```

```diff
   "scripts": {
-    "serve:ssr": "node dist/YOUR-APP-NAME/server/main.js",

> +    "serve:ssr": "node dist/YOUR-APP-NAME/server/server.mjs",

Reviewer: `MINOR:` Fresh apps with Angular >= 17 contains app's name in the script's name, e.g `"serve:ssr:YOUR-APP-NAME"`. If we want to be as close as the code generated by the latest Angular, we can consider this change.

Author: ```suggestion
+    "serve:ssr:YOUR-APP-NAME": "node dist/YOUR-APP-NAME/server/server.mjs",
```

- const mainModule = __non_webpack_require__.main;
- const moduleFilename = (mainModule && mainModule.- filename) || '';
- if (moduleFilename === __filename || moduleFilename.- includes('iisnode')) {
-   run();
- }

+ run();
```

6. In the very bottom of the file, remove the re-export of the path `./src/main.server`

```diff
- export * from './src/main.server';
```

This is the end of the migration guide for the Angular `application` builder.

## In `app.module.ts`, use new, non-deprecated Angular APIs

Unrelated to the migration to the new Angular `application` builder, please also update the `app.module.ts` file to use new, non-deprecated Angular APIs:

### `src/app/app.module.ts`

1. Remove the `HttpClientModule` from the `imports` array. 

```diff
  imports: [
-    HttpClientModule,
```

2. Add `provideHttpClient(withFetch(), withInterceptorsFromDi()),` to the `providers` array.

```diff
  providers: [
+   provideHttpClient(withFetch(), withInterceptorsFromDi()),
  ],
```

## For SSR projects, additionally:

### `src/app/app.module.ts`

Replace the item in the `imports` array `BrowserModule.withServerTransition({ appId: 'serverApp' }),` with just `BrowserModule`

> ```diff
>   imports: [
> -    BrowserModule.withServerTransition({ appId: 'serverApp' }),
> +    BrowserModule,
>   ],
> ```

Reviewer: ```suggestion
```diff
  imports: [
-    BrowserModule.withServerTransition({ appId: 'serverApp' }),
+    BrowserModule,
  ],
  /* ... */
  providers: [
    {
      provide: APP_ID,
      useValue: 'serverApp'
    }
  ]
```

```
We can propose an alternative to what was deprecated.
Note: It's a mandatory change when migrating to v19, because `withServerTransition` was removed from code in this version.

Author: I have problem to understand why would we provide the APP_ID. In fresh ng17 apps there is no such provider.
Does ng19 migration add this provider?

Reviewer: It's part of [Angular Migration Guide](https://angular.dev/update-guide?v=18.0-19.0&l=1) which should be adresses manually by customers.
![image](https://github.com/user-attachments/assets/01d9ecb9-62b2-47f5-a2bd-6463077acdeb)
If for some reasons customers need to use `appId` in their app, we can provide equivalent for previous `withServerTransition`
Another option is to not mention this change in our docs at all as they are already addressed in Angular Update Guide to v19 (as well as in [update guide to v16, when the method was deprecated](https://angular.dev/update-guide?v=15.0-16.0&l=3))

Author: I consider a need for providing any specific `APP_ID` an edge case in context of Spartacus, because we don't officially support having 2 adjacent Angular apps bootstraped in the same document. See Angular docs:

> 
> ```
> /**
>  * A DI token representing a string ID, used
>  * primarily for prefixing application attributes and CSS styles when
>  * {@link ViewEncapsulation#Emulated} is being used.
>  *
>  * The token is needed in cases when multiple applications are bootstrapped on a page
>  * (for example, using `bootstrapApplication` calls). In this case, ensure that those applications
>  * have different `APP_ID` value setup. For example:
>    (...)
>  ```
> ([Angular source](https://github.com/angular/angular/blob/aa45d44c5bee49638534895e8650dd3a9217e553/packages/core/src/application/application_tokens.ts#L12-L19))

That's why I'd advocate for just removing the deprecated part and not replacing it with any specific `APP_ID` provider.

Note: Angular's provides implicitly an OOTB default value of `APP_ID`: `"ng"`:
https://github.com/angular/angular/blob/aa45d44c5bee49638534895e8650dd3a9217e553/packages/core/src/application/application_tokens.ts#L42-L48
